Python/sysmodule.c

```diff
@@ -111,6 +112,266 @@ PySys_SetObject(const char *name, PyObject *v)
     }
 }
 
+int
+PySys_Audit(const char *event, const char *argFormat, ...)
+{
+    PyObject *eventName = NULL;
+    PyObject *eventArgs = NULL;
+    PyObject *hooks = NULL;
+    PyObject *hook = NULL;
+    int res = -1;
+    int dtrace = PyDTrace_AUDIT_ENABLED();
+
+    /* N format is inappropriate, because you do not know
+       whether the reference is consumed by the call.
+       Assert rather than exception for perf reasons */
+    assert(!argFormat || !strchr(argFormat, 'N'));
+
+    /* Early exit when no hooks are registered */
+    _Py_AuditHookEntry *e = _PyRuntime.audit_hook_head;
+    PyThreadState *ts = _PyThreadState_GET();
+    PyInterpreterState *is = ts ? ts->interp : NULL;
+    if (!e && (!is || !is->audit_hooks) && !dtrace) {
+        return 0;
+    }
+
+    PyObject *exc_type, *exc_value, *exc_tb;
+    if (ts) {
+        PyErr_Fetch(&exc_type, &exc_value, &exc_tb);
+    }
+
+    /* Initialize event args now */
+    if (argFormat && argFormat[0]) {
+        va_list args;
+        va_start(args, argFormat);
+        eventArgs = Py_VaBuildValue(argFormat, args);
+        if (eventArgs && !PyTuple_Check(eventArgs)) {
+            eventArgs = PyTuple_Pack(1, eventArgs);
+        }
+    } else {
+        eventArgs = PyTuple_New(0);
+    }
+    if (!eventArgs) {
+        goto exit;
+    }
+
+    /* Call global hooks */
+    for (; e; e = e->next) {
+        if (e->hookCFunction(event, eventArgs, e->userData) < 0) {
```

Comment:
This is where we may get into trouble with objects crossing the interpreter boundary, right?

Reply:
I don't think so. The hooks will be called in the same thread as an interpreter with the GIL held. ``e->userData`` is not designed to hold or store any ``PyObject*`` data.

(This should be documented, too.)

Reply:
Correct. The C function is not tied to a particular interpreter and will only be called in the context of the one where the arguments come from.

I'll improve the docs.